### ISSUE TITLE
Discussion on ido behavior

### ISSUE THREAD

**Issue by DarwinAwardWinner** [DarwinAwardWinner/ido-ubiquitous#106]
This is an issue for continuing the discussion begun at https://github.com/DarwinAwardWinner/ido-ubiquitous/commit/ec41a9739e9899518a39e4000e302d50b3b23e0c#commitcomment-15932721

**Comment by DarwinAwardWinner** [comment]
> > so pressing RET immediately should return the first item, as per the usual behavior of ido
>
> I think this is a mistake. Yes, pressing RET to select the first item is handy, but not when it's at the expense of the actual semantics of completing-read.

Pressing RET to select the default/first matching item is a core ido behavior. If people don't want that, they should probably use icomplete-mode or something else instead.

> On the other hand, doesn't C-j select the default (or an empty string), even without the overrides?

Pressing C-j is indeed the ido mechanism for selecting the currently entered text without performing any completion. For example, if you have a file called `abc.txt` and you want to create a new file called `abc` using `ido-find-file` you must use C-j to do so (or else use C-f to fall back to normal file completion). Similarly, C-j is the only way to enter an empty string in ordinary `ido-completing-read`. However, the purpose of C-j in ido is *not* to select the default item. Regardless, that is what you'd get for any function that uses the obsolete method of relying on an empty string as a signal to select the default. This is actually a terrible behavior, because it requires a user to know, for each command they use, whether it uses old-style default selection or not, so that they will know whether to hit RET or C-j to get the default. You might not agree that RET is the right key to always select the default or first item, but I hope you'll agree that having the key binding for this action change depending on which command is being executed is wrong behavior. In order to provide this consistency within ido completion, ido-ubiquitous needs to know which commands are using old-style default selection.

> Why are you saying it's obsolete? It's still documented.

Hmm. Now that I think about it, I suppose I've never actually seen it stated that returning an empty string when on null input is an obsolete behavior. I inferred this from the fact that I've only ever seen this feature used by code that was written before `completing-read` got a default argument in order to signal that the default was selected. I assumed that, since `completing-read` now has an argument to specify the default, and since all newer code that I've read uses the default argument instead of this obtuse method way of specifying the default, the feature must just be maintained for backward compatibility with that old code that still relies on it for default selection.

> Info-read-node-name is an example of using this semantics on purpose.

Looking at the code of `Info-read-node-name` (in Emacs 24.5.1), I don't see how it's "using" this feature. It looks more like it is working around the possibility of having an empty string returned even though `require-match` is `t`. 

> The sane solution is to avoid hacking around the current semantics, and push for better version(s) of `completing-read`.

I agree, and would support such a change, but that isn't the purpose of ido-ubiquitous. The purpose of ido-ubiquitous is to fit ido's square peg into `completing-read`'s round hole, not to carve a new square hole. As long as people want to use ido with the current incarnation of `completing-read`, it will require said hacking.

> Have you looked at company-capf, its usage of completion--capf-wrapper, and completion--capf-misbehave-funs?

I haven't personally used company mode, so I'm not familiar with the code, but with a brief glance it looks like `completion--capf-wrapper` implements a whitelist and blacklist of functions, conceptually similar to what ido-ubiquitous overrides are doing. I'm not sure which difference you want to call my attention to. Are you suggesting splitting the blacklist and whitelist into separate variables instead of having them together in a single "overrides" variable that includes both?

> As long as it breaks the completing-read-function API, it's a bug, and there's no need for you to support this package.

My goal is to avoid ever having ido-ubiquitous break any existing use of `completing-read`. I accept that ido completion is not a feature-complete replacement for `completing-read-default`, and so I avoid breakage avoiding ido in functions where it causes breakage. I think this is a reasonable approach, since a user experiencing breakage will likely think "this function worked fine until I enabled ido-ubiquitous, therefore ido-ubiquitous must be breaking it".

> Have you submitted a patch [to tmm] yet? The easiest fix seems to be to make it call completing-read-default directly.

That seems like an good fix. At the time I blacklisted tmm in ido-ubiquitous, I hadn't figured out what the source of the incompatibility was, so I just added an override and moved on to other things. When I later figured it out, it never occurred to me that it could so easily be fixed on the tmm side. I'll submit a patch for this when I get a chance.

> it would be much easier to maintain a blacklist of old completion functions that engage in obsolete practices like this, instead of having to whitelist the "compatible" completion functions.

At the time I wrote the code, my experience was that passing a function as the collection argument to `completing-read` was pretty rare, so a whitelist seemed reasonable, especially because it accomplishes my goal of never breaking existing uses of `completing-read`. However, I guess I wasn't *too* committed to a whitelist when I wrote this code, because I also implemented a user preference[1] to switch to a blacklist mode instead. In theory I could change that default in a future ido-ubiquitous release to get the behavior you're recommending, if I had reasonable confidence that it wouldn't break too many things. Maybe I should flip the switch on my own Emacs and see if I run into any issues.

[1] https://github.com/DarwinAwardWinner/ido-ubiquitous/blob/master/ido-ubiquitous.el#L575-L589

**Comment by DarwinAwardWinner** [comment]
Also, let me just say I appreciate you taking the time to have this discussion with me. I definitely don't know all the context and history behind the design choices of the completion system.

**Comment by dgutov** [comment]
> Pressing RET to select the default/first matching item is a core ido behavior. If people don't want that, they should probably use icomplete-mode or something else instead.

You can still have RET always pick the first item if there is a default, or if there's non-empty input. Anyway, I'm not married to a particular approach here, I'm just saying it should be solvable in one way or another without the hassle of a list of overrides that has to be maintained by hand.

> However, the purpose of C-j in ido is not to select the default item.

Why not fix that? `completing-read-default` won't allow the user to input an empty string if the default is specified. There's no reason for Ido to allow that either.

> This is actually a terrible behavior, because it requires a user to know, for each command they use, whether it uses old-style default selection or not, so that they will know whether to hit RET or C-j to get the default.

C-j will pick the default when there's no input. RET will just always select the first item in the list (you could arrange the default value to be there, but that's not 100% necessary).

> You might not agree that RET is the right key to always select the default or first item, but I hope you'll agree that having the key binding for this action change depending on which command is being executed is wrong behavior. In order to provide this consistency within ido completion, ido-ubiquitous needs to know which commands are using old-style default selection.

What's inconsistent there? If the user likes the first item, they can press RET. If they want empty input or the default, they can press C-j. Maybe the prompt should also say what the default is (especially when it's the empty string--i.e. no default).

> Looking at the code of Info-read-node-name (in Emacs 24.5.1), I don't see how it's "using" this feature. It looks more like it is working around the possibility of having an empty string returned even though require-match is t.

It seems you're right. But if it's not using that for anything useful, either there's an example of function that does use it, or you can comfortably ignore that possibility in ido-ubiquitious, and never return `""`.

> but with a brief glance it looks like completion--capf-wrapper implements a whitelist and blacklist of functions, conceptually similar to what ido-ubiquitous overrides are doing. I'm not sure which difference you want to call my attention to.

To the fact that no human being has to maintain the contents of those lists.

> I accept that ido completion is not a feature-complete replacement for completing-read-default, and so I avoid breakage avoiding ido in functions where it causes breakage. I think this is a reasonable approach, since a user experiencing breakage will likely think "this function worked fine until I enabled ido-ubiquitous, therefore ido-ubiquitous must be breaking it".

Since completing-read-function was introduced, I believe we can safely call certain clients, like tmm, to be misbehaving.

**Comment by dgutov** [comment]
> I definitely don't know all the context and history behind the design choices of the completion system.

Nor do I, really. If you have some in-depth questions, you'd be better off asking them on emacs-devel. This discussion could have attracted some extra comments as well, if it went on the mailing list.

>  In theory I could change that default in a future ido-ubiquitous release to get the behavior you're recommending, if I had reasonable confidence that it wouldn't break too many things.

It would break things. Like I mentioned in the first comment to the original discussion, the proper support for functional completion tables in Ido still needs to be implemented.

**Comment by DarwinAwardWinner** [comment]
Ok, here's a concrete example to demonstrate the problem with old-style defaults. Note that all of the below assumes you have loaded the `with-simulated-input` macro from [`ido-ubiquitous-test.el`](https://github.com/DarwinAwardWinner/ido-ubiquitous/blob/master/test/ido-ubiquitous-test.el) in this repository. I'm just using it here to simulate pressing RET programmatically. You'll also need to load ERT for the `should` macro.

### Normal `completing-read`

Everything works fine with `completing-read`. In both cases, pressing RET gives you the default, "b".

```elisp
;; New default method
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (completing-read-default "Select one (Default: b): " '("a" "b" "c") nil t nil nil "b"))))

;; Old default method
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (let* ((def "b")
           (prompt (format "Select one (Default: %s):" def))
           (answer (completing-read-default prompt '("a" "b" "c") nil t)))
      (if (string= answer "")
          def
        answer)))))
```

### Using `ido-completing-read`

The second assertion fails, because pressing RET gives you "a", the first item in the list, not "b", the intended default.

```elisp
;; New default method
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (ido-completing-read "Select one (Default: b): " '("a" "b" "c") nil t nil nil "b"))))

;; Old default method
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (let* ((def "b")
           (prompt (format "Select one (Default: %s):" def))
           (answer (ido-completing-read prompt '("a" "b" "c") nil t)))
      (if (string= answer "")
          def
        answer)))))
```

This result is highly unintuitive for the user since the prompt implies that the default is "b", and the user has no way to tell which of the above cases in in effect short of tracking down and looking at the code that is being run.

### Using `completing-read-ido-ubiquitous` and setting overrides correctly

Now everything works again. We always get "b" when pressing RET, which is consistent with what the prompt tells us.

```elisp
;; New default method
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (completing-read-ido-ubiquitous "Select one (Default: b): " '("a" "b" "c") nil t nil nil "b"))))

;; Old default method
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (let* ((def "b")
           (prompt (format "Select one (Default: %s):" def))
           (answer (ido-ubiquitous-with-override 'enable-old
                     (completing-read-ido-ubiquitous prompt '("a" "b" "c") nil t))))
      (if (string= answer "")
          def
        answer)))))
```

### Using `completing-read-ido-ubiquitous` without overrides overrides correctly

This just demonstrates that `completing-read-ido-ubiquitous` only works because of the override. The code below will produce the same result as `ido-completing-read`: it will return "a" and fail the assertion.

```elisp
(should
 (string=
  "b"
  (with-simulated-input "RET"
    (let* ((def "b")
           (prompt (format "Select one (Default: %s):" def))
           (answer (completing-read-ido-ubiquitous prompt '("a" "b" "c") nil t)))
      (if (string= answer "")
          def
        answer)))))
```

**Comment by dgutov** [comment]
What makes you think this usage is popular enough to care about? Even if it is:

* It should be easy to fix at the call sites to use the new convention. I'm fairly sure `completing-read` has supported the DEF argument for years.

* If the example is faithful to reality, it should be easy to detect dynamically: `(default:` is in PROMPT, but DEF is nil.

**Comment by DarwinAwardWinner** [comment]
I probably wouldn't have implemented support for old-style defaults if I hadn't encountered them in the wild on a fairly regular basis. If I recall correctly, it drove me crazy that ido-ubiquitous sometimes wouldn't select the indicated default and sometimes would, and it took me several months of just putting up with it before I finally figured out why. The default set of overrides documents all the cases that I've found or been informed of to date. Unfortunately, I don't have the time right now to find and fix all the uses of the old convention in the Emacs codebase. (I'm sure my override list is not exhaustive.) As for parsing the default out of the prompt, I don't think that every use of the old convention uses the same syntax for including the default in the prompt. The intention of my example was to show a representative example of how the default is included in an arbitrary fashion into the prompt, but not made directly available to `completing-read`. I don't think trying to parse the default out of a prompt that is essentially freeform text would be an improvement on the current manually-curated overrride list.

Bottom line, I don't want to break things that work right now in hopes that it will encourage them to get fixed on the other end, but any ido-ubiquitous user interested in doing so can simply set `ido-ubiquitous-command-overrides` and `ido-ubiquitous-function-overrides` to nil. Your original point was about supporting functional completion tables in ido. I agree this would be great, and would eliminate one reason for the override system. Rewriting all legacy code to use the DEF argument would eliminate the other reason. If both of those could be accomplished, then it would probably be possible to remove the override system entirely from ido-ubiquitous, or even just retire ido-ubiquitous entirely, in favor of `(setq completing-read-function 'ido-completing-read)`.
